framework/harness/harness.go

```diff
@@ -0,0 +1,138 @@
+	startupOutput io.Writer,
+) (*TestHarness, error) {
+	if debugLogger == nil {
+		debugLogger = framework.NullLogger()
+	}
+
+	externalBaseURL := fmt.Sprintf("http://%s:%d", testHarnessExternalHostname, testHarnessPort)
+
+	h := &TestHarness{
+		testServiceBaseURL:         testServiceBaseURL,
+		testHarnessExternalBaseURL: externalBaseURL,
+		mockEndpoints:              newMockEndpointsManager(externalBaseURL, debugLogger),
+		logger:                     debugLogger,
+	}
+
+	testServiceInfo, err := queryTestServiceInfo(testServiceBaseURL, statusQueryTimeout, startupOutput)
+	if err != nil {
+		return nil, err
+	}
+	h.testServiceInfo = testServiceInfo
+
+	if err = startServer(testHarnessPort, http.HandlerFunc(h.serveHTTP)); err != nil {
+		return nil, err
+	}
+
+	return h, nil
+}
+
+// TestServiceInfo returns the initial status information received from the test service.
+func (h *TestHarness) TestServiceInfo() TestServiceInfo {
+	return h.testServiceInfo
+}
+
+// NewMockEndpoint adds a new endpoint that can receive requests.
+//
+// The specified handler will be called for all incoming requests to the endpoint's
+// base URL or any subpath of it. For instance, if the generated base URL (as reported
+// by MockEndpoint.BaseURL()) is http://localhost:8111/endpoints/3, then it can also
+// receive requests to http://localhost:8111/endpoints/3/some/subpath.
+//
+// When the handler is called, the test harness rewrites the request URL first so that
+// the handler sees only the subpath. It also attaches a Context to the request whose
+// Done channel will be closed if Close is called on the endpoint.
+func (h *TestHarness) NewMockEndpoint(
+	handler http.Handler,
+	contextFn func(context.Context) context.Context,
+	logger framework.Logger,
+) *MockEndpoint {
+	if logger == nil {
+		logger = h.logger
+	}
+	return h.mockEndpoints.newMockEndpoint(handler, contextFn, logger)
+}
+
+func (h *TestHarness) serveHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "HEAD" {
+		w.WriteHeader(200) // we use this to test whether our own listener is active yet
+		return
+	}
+	h.mockEndpoints.serveHTTP(w, r)
+}
+
+func startServer(port int, handler http.Handler) error {
+	server := &http.Server{
+		Addr: fmt.Sprintf(":%d", port),
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if r.Method == "HEAD" {
+				w.WriteHeader(200)
+				return
+			}
+			handler.ServeHTTP(w, r)
+		}),
+	}
+	go func() {
+		if err := server.ListenAndServe(); err != nil {
+			panic(err)
+		}
+	}()
+
+	// Wait till the server is definitely listening for requests before we run any tests
+	deadline := time.NewTimer(httpListenerTimeout)
+	defer deadline.Stop()
+	ticker := time.NewTicker(time.Millisecond * 10)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-deadline.C:
+			return fmt.Errorf("could not detect own listener at %s", server.Addr)
+		case <-ticker.C:
+			resp, err := http.DefaultClient.Head(fmt.Sprintf("http://localhost:%d", port))
```

Comment:
"It may not make progress for a couple iterations" is not a problem; we're not concerned about wasting a few CPU cycles during the startup of this process.

The goal here is not to make the minimum possible number of checks, but to detect success as soon as possible so we don't delay for an unnecessarily long time. If you changed the `time.NewTicker` parameter to half a second, what would happen is that the first listener check would fail (very likely, since `NewTicker` sends a tick _immediately_) and then the app would wait half a second. Every startup, or nearly every startup, would take at least half a second.